### PR TITLE
Inject SMTLIB2 queries and responses via standard-json-io.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
Unfortunately, this is quite hard to test, since z3 is either compiled in or not, and if z3 is compiled in, we cannot use the smtlib2 interface.

File: libsolidity/interface/CompilerStack.h [hunk 1]
```diff
@@ -130,6 +130,13 @@ class CompilerStack: boost::noncopyable
 	/// @returns true if a source object by the name already existed and was replaced.
 	bool addSource(std::string const& _name, std::string const& _content, bool _isLibrary = false);
 
+	/// Adds a response to an SMTLib2 query (identified by the hash of the query input).
+	void addSMTLib2Response(h256 const& _hash, std::string const& _response) { m_smtlib2Responses[_hash] = _response; }
+
+	/// @returns a list of unhandled queries to the SMT solver (has to be supplied in a second run
+	/// by calling @a addSMTLib2Response.
+	std::vector<std::string> const& unhandledSMTQueries() const { return m_unhandledSMTLib2Queries; }
```
Comment (axic) [inline libsolidity/interface/CompilerStack.h]: Probably these two should match, ie `unhandledSMTLib2Queries`.

Also, it is nice to pair these two, but at the moment the functions are ordered in this file roughly by the order one can call them. That means this particular one should move after `compile` (well, after analyze, but parse/analyze/compiler are kept together).

File: libsolidity/interface/CompilerStack.h [hunk 2]
```diff
@@ -130,6 +130,13 @@ class CompilerStack: boost::noncopyable
 	/// @returns true if a source object by the name already existed and was replaced.
 	bool addSource(std::string const& _name, std::string const& _content, bool _isLibrary = false);
 
+	/// Adds a response to an SMTLib2 query (identified by the hash of the query input).
```
Comment (axic) [inline libsolidity/interface/CompilerStack.h:156]: So the hash is the hash of one of the queries returned by a previous run?

I think we should raise an error if not all the response hashes are used because that means the run was different.

File: libsolidity/interface/CompilerStack.h (hunk 1 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.h]: I thought about this, too, but it is actually much more similar to `errors()` and that is also rather to the start.

File: libsolidity/interface/CompilerStack.h (hunk 2 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.h:156]: Yes. It might happen that the second run requests new queries (because e.g. something only makes sense to ask depending on the result of a previous question), but probably all of them should be used. What kind of error would that be? A regular error supplied to the user?

Comment (axic) [reply libsolidity/interface/CompilerStack.h:156]: I'm not sure how feasible it is, but yeah even if not an error, but at least a warning stating that not all supplied SMT responses were processed.

File: libsolidity/interface/CompilerStack.h (hunk 1 above)
Comment (axic) [reply libsolidity/interface/CompilerStack.h]: Hm, I'd say `errors()` is an oversight in that case :)

File: libsolidity/interface/CompilerStack.h (hunk 2 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.h:156]: Hm, this is quite hard to implement, actually, because smtlib2 does not have access to the error object and we would have to track that across contracts and thus across smtlib2 instances. I will add an error, though, if those queries were given, but we are using z3 directly.

File: libsolidity/interface/StandardCompiler.cpp [hunk 3]
```diff
@@ -255,6 +255,27 @@ Json::Value StandardCompiler::compileInternal(Json::Value const& _input)
 			return formatFatalError("JSONError", "Invalid input source specified.");
 	}
 
+	Json::Value const& auxInputs = _input["auxiliaryInput"];
+	if (!!auxInputs)
+	{
+		Json::Value const& smtlib2Responses = auxInputs["smtlib2"];
+		if (!!smtlib2Responses)
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp:326]: ~~Why is this needed?~~

**Review by axic — CHANGES_REQUESTED** [review]
Need to document this in the jsonio section.

**Comment by leonardoalt** [comment]
I rebased this branch and found a few problems with the `SMTLib2Interface`. Working on them.

**Comment by axic** [comment]
@leonardoalt can you pull out the smtlib fixes and can we merge them already?

**Comment by axic** [comment]
Rebased and added some tiny fixes (will need to squash).

**Comment by axic** [comment]
Needs "standard json io" documentation update.

File: libsolidity/interface/StandardCompiler.cpp [hunk 4]
```diff
@@ -489,6 +510,10 @@ Json::Value StandardCompiler::compileInternal(Json::Value const& _input)
 	if (errors.size() > 0)
 		output["errors"] = errors;
 
+	if (!m_compilerStack.unhandledSMTLib2Queries().empty())
+		for (string const& query: m_compilerStack.unhandledSMTLib2Queries())
+			output["auxiliaryInputRequested"]["smtlib2"]["0x" + keccak256(query).hex()] = query;
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp]: I think this should be named `auxiliaryOutput`, potentially also `smtlib2queries` may be better than `smtlib2`.

File: libsolidity/interface/StandardCompiler.cpp [hunk 5]
```diff
@@ -316,6 +316,27 @@ Json::Value StandardCompiler::compileInternal(Json::Value const& _input)
 			return formatFatalError("JSONError", "Invalid input source specified.");
 	}
 
+	Json::Value const& auxInputs = _input["auxiliaryInput"];
+	if (!!auxInputs)
+	{
+		Json::Value const& smtlib2Responses = auxInputs["smtlib2"];
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp]: Likewise we could consider naming this `smtlib2responses` or similar.

File: libsolidity/interface/StandardCompiler.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/interface/StandardCompiler.cpp]: I like `smtlib2queries` better, `auxiliaryOutput` sounds vague.

File: libsolidity/interface/StandardCompiler.cpp (hunk 5 above)
Comment (leonardoalt) [reply libsolidity/interface/StandardCompiler.cpp]: Yes, sounds good.

**Comment by leonardoalt** [comment]
@axic You mean the fixes related to repeated declarations?

**Comment by axic** [comment]
> You mean the fixes related to repeated declarations?

Yes, the commit you've made.

**Comment by leonardoalt** [comment]
Will apply that commit on develop after the SMTPortfolio PR gets merged, since it will need adjustments.

**Comment by axic** [comment]
@leonardoalt portfolio is merged, would be nice to pull that commit and merge next in a new PR.

**Comment by axic** [comment]
@leonardoalt as #4603 is merged, can you rebase this please?

**Comment by leonardoalt** [comment]
@axic Rebased

File: libsolidity/formal/SMTLib2Interface.h [hunk 6]
```diff
@@ -77,6 +77,8 @@ class SMTLib2Interface: public SolverInterface, public boost::noncopyable
 	std::set<std::string> m_functions;
 
 	std::vector<std::string> m_unhandledQueries;
+	std::set<std::string> m_constants;
+	std::set<std::string> m_functions;
```
Comment (axic) [inline libsolidity/formal/SMTLib2Interface.h]: Duplicates.

File: libsolidity/interface/StandardCompiler.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/interface/StandardCompiler.cpp]: I actually meant `auxiliaryInputRequest.smtlib2queries` 😉

File: libsolidity/interface/StandardCompiler.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/interface/StandardCompiler.cpp]: Likewise: `auxiliaryInput.smtlib2responses`.

File: libsolidity/formal/SMTLib2Interface.h [hunk 7]
```diff
@@ -67,10 +71,12 @@ class SMTLib2Interface: public SolverInterface, public boost::noncopyable
 	/// Communicates with the solver via the callback. Throws SMTSolverError on error.
 	std::string querySolver(std::string const& _input);
 
-	ReadCallback::Callback m_queryCallback;
+	std::map<h256, std::string> const& m_smtlib2Responses;
 	std::vector<std::string> m_accumulatedOutput;
 	std::set<std::string> m_constants;
 	std::set<std::string> m_functions;
+
+	std::vector<std::string> m_unhandledQueries;
```
Comment (axic) [inline libsolidity/formal/SMTLib2Interface.h:80]: I guess we have to name this like `m_smtlib2UnhandledQueries`

File: libsolidity/formal/SMTPortfolio.cpp [hunk 8]
```diff
@@ -41,9 +41,15 @@ SMTPortfolio::SMTPortfolio(ReadCallback::Callback const& _readCallback)
 	m_solvers.emplace_back(make_shared<smt::CVC4Interface>());
 #endif
 #if !defined (HAVE_Z3) && !defined (HAVE_CVC4)
-	m_solvers.emplace_back(make_shared<smt::SMTLib2Interface>(_readCallback)),
+	m_solvers.emplace_back(make_shared<smt::SMTLib2Interface>(_smtlib2Responses));
+#else
+	if (!_smtlib2Responses.empty())
+		m_errorReporter.warning(
+			"Query responses for smtlib2 were given in the auxiliary input, "
```
Comment (axic) [inline libsolidity/formal/SMTPortfolio.cpp]: Actually what is the official styling, smtlib2 or SMTLib2?

File: libsolidity/formal/SMTLib2Interface.h (hunk 7 above)
Comment (axic) [reply libsolidity/formal/SMTLib2Interface.h:80]: Never mind, this is in `SMTLib2Interface`. In this case then we could rename `m_stlib2Responses` to `m_externalResponses` or just `m_queryResponses`?

Comment (leonardoalt) [reply libsolidity/formal/SMTLib2Interface.h:80]: `m_smtlib2Responses` is above, this is something else, right?

File: libsolidity/formal/SMTPortfolio.cpp (hunk 8 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTPortfolio.cpp]: For messages like this maybe it is better to go with SMT-LIB2

**Comment by leonardoalt** [comment]
Fixed some suggestions.
Will need a good squash when it's ready ;)

File: libsolidity/formal/SMTLib2Interface.h (hunk 7 above)
Comment (axic) [reply libsolidity/formal/SMTLib2Interface.h:80]: It is just confusing that these exists on `CompilerStack` and here with inconsistent names.

For this file I'd suggest to group the two members together and name them:
- `m_queryResponses`
- `m_unhandledQueries` or `m_queryRequests`

File: libsolidity/formal/SMTPortfolio.cpp [hunk 9]
```diff
@@ -41,9 +41,15 @@ SMTPortfolio::SMTPortfolio(ReadCallback::Callback const& _readCallback)
 	m_solvers.emplace_back(make_shared<smt::CVC4Interface>());
 #endif
 #if !defined (HAVE_Z3) && !defined (HAVE_CVC4)
-	m_solvers.emplace_back(make_shared<smt::SMTLib2Interface>(_readCallback)),
+	m_solvers.emplace_back(make_shared<smt::SMTLib2Interface>(_smtlib2Responses));
+#else
+	if (!_smtlib2Responses.empty())
+		m_errorReporter.warning(
+			"SMT-LIB2 query responses were given in the auxiliary input, "
+			"but this Solidity binary uses an SMT solver directly."
```
Comment (axic) [inline libsolidity/formal/SMTPortfolio.cpp]: Should add: these responses will be ignored. Consider disabling Z3/CVC4 at compilation time in order to use SMT-LIB2 responses.

File: libsolidity/interface/StandardCompiler.cpp [hunk 10]
```diff
@@ -316,6 +316,23 @@ Json::Value StandardCompiler::compileInternal(Json::Value const& _input)
 			return formatFatalError("JSONError", "Invalid input source specified.");
 	}
 
+	Json::Value const& smtlib2Responses = _input["auxiliaryInput.smtlib2responses"];
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp]: Sorry I keep confusing you. I was ok with the two-levels of JSON, just suggested a different name for both levels.

Comment (leonardoalt) [reply libsolidity/interface/StandardCompiler.cpp]: Haha ok

**Comment by chriseth** [comment]
@leonardoalt what is the plan for this PR?

**Comment by leonardoalt** [comment]
Still need to make SMT queries work in JS.

**Comment by axic** [comment]
@leonardoalt don't you think this could be merged independently?

Can we add a test case for this?

**Comment by leonardoalt** [comment]
@axic Yes, let's do that. Just rebased.

The current code says the `_smtlib2Responses` will be ignore if Z3 or CVC4 are available, but they are forwarded to `SMTLib2Interface` anyway. I'd say we keep `SMTLib2Interface` always enabled and remove the error message. What do you think?

**Comment by leonardoalt** [comment]
@axic Created a unit test for smt via standard json. It does not run the solver on the queries: it expects the Solidity code and a JSON file with the answers for the generated queries.

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+	{
+		Json::Value const& smtlib = auxInputs[_smtlib];
+		if (!!smtlib)
+			for (auto const& hashString: smtlib.getMemberNames())
+				hashes.push_back(hashString);
+	}
+	return hashes;
+}
+
+string SMTCheckerTest::buildJsonContent(std::string const& _extra)
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: I think this should take a `Json::Value` for `auxiliaryContent`, which can be a null type. This would greatly simplify the caller.

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+		input += ", " + _extra;
+	input += "}";
+
+	return input;
+}
+
+Json::Value SMTCheckerTest::compile(string const& _input)
+{
+	StandardCompiler compiler;
+	string output = compiler.compile(_input);
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: Also note that `compile` has an overload which takes `Json::Value` and returns `Json::Value`.

Comment (leonardoalt) [reply test/libsolidity/SMTCheckerJSONTest.cpp]: Ah nice

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	Json::Value ret;
+	jsonParse(output, ret);
+	return ret;
+}
+
+void SMTCheckerTest::parseJsonFile(string const& _jsonFilename)
+{
+	ifstream file(_jsonFilename);
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: I think we have a helper in `devcore` for reading a file: `readFileAsString` in `CommonIO.h`

Probably makes sense putting a helper there which reads a file and parses it as json.

Comment (leonardoalt) [reply test/libsolidity/SMTCheckerJSONTest.cpp]: What about putting the helper in `JSON.h/cpp`?

**Comment by leonardoalt** [comment]
@axic Updated

**Comment by axic** [comment]
@leonardoalt why haven't we merged this yet?

**Comment by axic** [comment]
@leonardoalt rebased this, but I think we need to squash some commits.

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+#include <memory>
+#include <stdexcept>
+#include <sstream>
+
+using namespace dev;
+using namespace solidity;
+using namespace dev::solidity::test;
+using namespace dev::solidity::test::formatting;
+using namespace std;
+namespace fs = boost::filesystem;
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: This doesn't seem to be used.

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+Json::Value SMTCheckerTest::buildJson()
+{
+	string language = "\"language\": \"Solidity\"";
+	string sourceName = "\"A\"";
+	string sourceContent = "\"" + m_source + "\"";
+	string sourceObj = "{ \"content\": " + sourceContent + "}";
+	string sources = " \"sources\": { " + sourceName + ": " + sourceObj + "}";
+	string input = "{" + language + ", " + sources + "}";
+	Json::Value source;
+	jsonParse(input, source);
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: Should have `BOOST_REQUIRE` around `jsonParse`.

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+using namespace dev::solidity::test::formatting;
+using namespace std;
+namespace fs = boost::filesystem;
+using namespace boost::unit_test;
+
+SMTCheckerTest::SMTCheckerTest(string const& _filename)
+: SyntaxTest(_filename)
+{
+	if (!boost::algorithm::ends_with(_filename, ".sol"))
+		BOOST_THROW_EXCEPTION(runtime_error("Invalid test contract file name: \"" + _filename + "\"."));
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: Why not `BOOST_REQUIRE_MSG(ends_with(), "message..")` ?

**Comment by leonardoalt** [comment]
@axic updated after comments

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+	StandardCompiler compiler;
+
+	Json::Value input = buildJson();
+	Json::Value result = compiler.compile(input);
+
+	vector<string> outHashes = hashesFromJson(result, "auxiliaryInputRequested", "smtlib2queries");
+	string auxInput("auxiliaryInput");
+	vector<string> inHashes = hashesFromJson(m_smtResponses, auxInput, "smtlib2responses");
+
+	BOOST_CHECK(outHashes == inHashes);
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: `CHECK_EQUAL(outHashes, inHashes)` ?

File: test/libsolidity/SMTCheckerJSONTest.cpp [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+	BOOST_CHECK(m_smtResponses.isMember(auxInput));
+	input[auxInput] = m_smtResponses[auxInput];
+
+	Json::Value endResult = compiler.compile(input);
+
+	BOOST_CHECK(endResult.isMember("errors"));
+	Json::Value const& errors = endResult["errors"];
+	for (auto const& err: errors)
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: Should use `error`.

File: test/libsolidity/SMTCheckerJSONTest.h [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+namespace solidity
+{
+namespace test
+{
+
+class SMTCheckerTest: public SyntaxTest
+{
+public:
+	static std::unique_ptr<TestCase> create(std::string const& _filename)
+	{ return std::unique_ptr<TestCase>(new SMTCheckerTest(_filename)); }
```
Comment (axic) [inline test/libsolidity/SMTCheckerJSONTest.h]: This formatting seems to be off.

File: test/libsolidity/SMTCheckerJSONTest.cpp (hunk 17 above)
Comment (leonardoalt) [reply test/libsolidity/SMTCheckerJSONTest.cpp]: `BOOST_CHECK_EQUAL` won't compile with vectors

Comment (axic) [reply test/libsolidity/SMTCheckerJSONTest.cpp]: Ok!

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
Need to be documented, created an issue #5483

**Comment by axic** [comment]
Locally I get the same queries as Circleci:
```
	"auxiliaryInputRequested" : 
	{
		"smtlib2queries" : 
		{
			"0x0426cd198d1e7123a28ffac2b759a666b86508ad046babf5166500dd6d8ed308" : "(set-option :produce-models true)\n(set-logic QF_UFLIA)\n(declare-fun |x_4_0| () Int)\n(assert (>= x_4_0 0))\n(assert (<= x_4_0 115792089237316195423570985008687907853269984665640564039457584007913129639935))\n(declare-fun |expr_8_0| () Int)\n(assert (= expr_8_0 x_4_0))\n(declare-fun |expr_9_0| () Int)\n(assert (= expr_9_0 0))\n(declare-fun |expr_10_0| () Bool)\n(assert (= expr_10_0 (> expr_8_0 expr_9_0)))\n(assert (or (not true) expr_10_0))\n(declare-fun |expr_14_0| () Int)\n(assert (= expr_14_0 x_4_0))\n(declare-fun |expr_15_0| () Int)\n(assert (= expr_15_0 100))\n(declare-fun |expr_16_0| () Bool)\n(assert (= expr_16_0 (> expr_14_0 expr_15_0)))\n(assert (or (not true) expr_16_0))\n(declare-fun |expr_20_0| () Int)\n(assert (= expr_20_0 x_4_0))\n(declare-fun |expr_21_0| () Int)\n(assert (= expr_21_0 0))\n(declare-fun |expr_22_0| () Bool)\n(assert (= expr_22_0 (>= expr_20_0 expr_21_0)))\n\n(assert (and true (not expr_22_0)))\n(declare-const |EVALEXPR_0| Int)\n(assert (= |EVALEXPR_0| x_4_0))\n(check-sat)\n(get-value (|EVALEXPR_0| ))\n",
			"0xa51ca41ae407f5a727f27101cbc079834743cc8955f9f585582034ca634953f6" : "(set-option :produce-models true)\n(set-logic QF_UFLIA)\n(declare-fun |x_4_0| () Int)\n(assert (>= x_4_0 0))\n(assert (<= x_4_0 115792089237316195423570985008687907853269984665640564039457584007913129639935))\n(declare-fun |expr_8_0| () Int)\n(assert (= expr_8_0 x_4_0))\n(declare-fun |expr_9_0| () Int)\n(assert (= expr_9_0 0))\n(declare-fun |expr_10_0| () Bool)\n(assert (= expr_10_0 (> expr_8_0 expr_9_0)))\n(assert (or (not true) expr_10_0))\n(declare-fun |expr_14_0| () Int)\n(assert (= expr_14_0 x_4_0))\n(declare-fun |expr_15_0| () Int)\n(assert (= expr_15_0 100))\n(declare-fun |expr_16_0| () Bool)\n(assert (= expr_16_0 (> expr_14_0 expr_15_0)))\n\n(assert (and true (not expr_16_0)))\n(declare-const |EVALEXPR_0| Int)\n(assert (= |EVALEXPR_0| x_4_0))\n(check-sat)\n(get-value (|EVALEXPR_0| ))\n",
			"0xe9477f683ff20aa57fcb08682150f86c5917e1d4c0686b278ab9b73446d0682c" : "(set-option :produce-models true)\n(set-logic QF_UFLIA)\n(declare-fun |x_4_0| () Int)\n(assert (>= x_4_0 0))\n(assert (<= x_4_0 115792089237316195423570985008687907853269984665640564039457584007913129639935))\n(declare-fun |expr_8_0| () Int)\n(assert (= expr_8_0 x_4_0))\n(declare-fun |expr_9_0| () Int)\n(assert (= expr_9_0 0))\n(declare-fun |expr_10_0| () Bool)\n(assert (= expr_10_0 (> expr_8_0 expr_9_0)))\n\n(assert (and true (not expr_10_0)))\n(declare-const |EVALEXPR_0| Int)\n(assert (= |EVALEXPR_0| x_4_0))\n(check-sat)\n(get-value (|EVALEXPR_0| ))\n"
		}
	},
```

@leonardoalt what queries are you getting?

**Comment by leonardoalt** [comment]
@axic In the local test run I get the same hashes, so I've updated the tests with those, should be enough for the PR.

But when I run
`./solc --allow-paths . --standard-json < multi_content.json`
where `multi_content.json` is
```
{
    "language" : "Solidity",
    "sources" :
    {
        "A" :
        {
            "content" : "pragma solidity >=0.0;\npragma solidity ^0.5.0;\npragma experimental SMTChecker;\n\ncontract C\n{\n\tfunction f(uint x) public pure {\n\t\tassert(x > 0);\n\t\tassert(x > 100);\n\t\tassert(x >= 0);\n\t}\n}\n"
        }
    }
}
```
I get different hashes.
If I use `url` instead of `content` in the JSON file, I get more different hashes.

Can you run this locally and see which hashes you get?

**Comment by axic** [comment]
I get the very same original hashes you got.

**Comment by axic** [comment]
Your JSON above does not match with the one generated:
```
{
    "language" : "Solidity",
    "sources" : 
    {
        "A" : 
        {
            "content" : "pragma solidity >=0.0;\npragma experimental SMTChecker;\n\ncontract C\n{\n\tfunction f(uint x) public pure {\n\t\tassert(x > 0);\n\t\tassert(x > 100);\n\t\tassert(x >= 0);\n\t}\n}\n"
        }
    }
}
```

You have an extra version pragma and since one of the SMT queries has a line number in it, you get a different hash.

**Comment by axic** [comment]
Squashed the last 3 commits.

**Review by axic — APPROVED** [review]
